### DISCUSSION TITLE
Github actions and required checks in a monorepo (using paths to limit execution)

### DISCUSSION THREAD

**Discussion by jondkoon** [community/community#26251] [Actions]
<p>I have a workflow defined like this:</p>
<pre><code>name: Api Lint

on:
  push:
    branches:
    - master
    paths:
    - api/**
  pull_request:
    paths:
    - api/**

jobs:
  api-lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v1
    - name: Use Node.js 12.x
      uses: actions/setup-node@v1
      with:
        node-version: 12.x
    - run: yarn install
    - name: api lint
      working-directory: ./api
      env:
        CI: true
      run: |
        yarn lint
</code></pre>
<p>You’ll notice this job only runs when the pull_request changes a file under the api directory. I’d like to make this a required check using the “Require status checks to pass before merging” setting, but when I make a change that does not touch the api directory the check does not run, so does not pass and I can’t merge my change with that enforcement in place.</p>
<p>Any ideas? Thanks!</p>

**Comment by SamChou19815** [comment, marked as answer]
<p>I used a hack to achieve what you want:<br>
I created another workflow with dummy jobs that have the same names with those required jobs that are not always triggered.<br>
In case the job is triggered and failed, it still still be push blocking according to required check. If it’s not triggered, then the dummy job will always succeed and unblocks the push.<br>
I hope GitHub can build some native solution for this.</p>

**Reply by TWiStErRob** [reply]
When using this workaround, beware of [the warnings in the docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks) (one at top in red, and one at the end after the example):

> It is recommended that you do not use path filtering (as shown in the previous example), or branch filtering, in a workflow that has been configured to be required. For more information, see "[Required workflows](https://docs.github.com/en/actions/using-workflows/required-workflows)."

**Reply by itxzainoooo332b** [reply]
That’s a clever workaround! It’s a smart way to handle conditional job requirements while still satisfying GitHub’s required checks. Totally agree a native solution from GitHub would make this much cleaner and[ more reliable. (https://thesnapchatmodapk.com/)

**Comment by jondkoon** [comment]
<p>That’s a wild hack, lol. Would you mind sharing what your dummy job looks like? Thanks! For example, does it still start the container and have all of that overhead?</p>

**Comment by SamChou19815** [comment]
<p>Something like</p>
<pre><code>name: Dummy
on: push

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - name: Do nothing
        run: exit 0
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Do nothing
        run: exit 0
</code></pre>

**Comment by hmehta** [comment]
<p>I would not really call such hack a sane solution, but would rather see native implementation skipping checks for pull requests that do not touch paths defined for the action pull request event.</p>

**Comment by solarmosaic-kflorence** [comment]
<p>Yes, can we re-open this so Github can make a real solution?</p>
<p>EDIT: created <a href="https://github.community/t/feature-request-conditional-required-checks/16761">https://github.community/t5/GitHub-Actions/Feature-request-conditional-required-checks/m-p/36938#M2735</a></p>

**Reply by danielnginx** [reply]
@solarmosaic-kflorence did they close the issue? I cannot find it.

**Comment by scherroman** [comment]
<p>It may be dicey, but as it stands <a class="mention" href="/samchou19815">@samchou19815</a>’s method of always triggering the dummy jobs that are named the same as the real jobs appears necessary to enable requiring different checks for different code paths. For example, requiring some tests to pass for changes to one service, but a different set of tests or no tests at all for another service or updates to the docs.</p>
<p>When there are multiple checks registered on the pull request with the same name, they all have to pass to fulfill the requirement. Also, the dummy job and the real job both start up at about the same time, so there’s no real window where the dummy has a chance to finish and allow merging before the real job starts up.</p>
<p>I call my dummy “automerge”</p>
<pre><code>name: Automerge
on:
  pull_request:
    branches:
      - master
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Exit with success
        run: exit 0
</code></pre>

**Comment by cscleison** [comment]
<p>For anyone wondering, you can do this with the <code>paths-filter</code> action. It runs a job to check which path changed (checkout not needed for this job, which is a plus), then the following jobs will conditionally run or skip. Then you can mark them all as required.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/dorny/paths-filter#conditional-execution">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/dorny/paths-filter#conditional-execution" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/4950497/181096437-96e26043-5142-4b3c-bca5-7ecb41ab5168.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/c/6cc31732ebd46f657490b651cf24acbd2f30f333_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/6/c/6cc31732ebd46f657490b651cf24acbd2f30f333_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/6/c/6cc31732ebd46f657490b651cf24acbd2f30f333.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/c/6cc31732ebd46f657490b651cf24acbd2f30f333_2_10x10.png"></div>

<h3><a href="https://github.com/dorny/paths-filter#conditional-execution" target="_blank" rel="noopener nofollow ugc">dorny/paths-filter</a></h3>

  <p>Conditionally run actions based on files modified by PR, feature branch or pushed commits - dorny/paths-filter</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by rytswd** [comment]
<p>I understand it’s quite an old ticket, but we have actually created a new GitHub Action for this use case.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/upsidr/merge-gatekeeper">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/upsidr/merge-gatekeeper" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/23435099/181096439-6914be17-35f2-4638-b153-034c08e4c482.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4ccc73aaddd4918b46a74e326a9c1c446af970ea_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4ccc73aaddd4918b46a74e326a9c1c446af970ea_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/c/4ccc73aaddd4918b46a74e326a9c1c446af970ea.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4ccc73aaddd4918b46a74e326a9c1c446af970ea_2_10x10.png"></div>

<h3><a href="https://github.com/upsidr/merge-gatekeeper" target="_blank" rel="noopener nofollow ugc">GitHub - upsidr/merge-gatekeeper: Get better merge control</a></h3>

  <p>Get better merge control. Contribute to upsidr/merge-gatekeeper development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Merge Gatekeeper is still in its early days, but it comes with a simple mindset that all jobs that ran should be in a success state. Merge Gatekeeper’s job is to wait and check other job status, and only becomes successful when all other jobs ran to completion and in a success state.</p>
<p>With that simple logic, you will only need to add a new Action setup using Merge Gatekeeper to your monorepo, and set it as the only required job - no other jobs need to be marked as required. As mentioned, the assumption is that no job should be in failed state, but this works nicely without any complex Action definitions.</p>
<p>If you need more granular control and ignore some failed state, it may be better handled with the solution <a class="mention" href="/cscleison">@cscleison</a> has referenced above, which looks great for some paths based branching :+1:</p>

**Comment by russell-dot-js** [comment]
<p>This worked great, thank you</p>

**Comment by m3h0w** [comment]
Anything we can do to reopen the conversation about this? It's quite confusing that the hacky solution is marked as an answer and the only other options are actions that overwrite the native GitHub logic.

It should be natively possible to configure the checks in a way that actions that don't run for a specific branch/PR. I.e. don't allow merging when an action fails instead of allowing merging if all actions succeed.

**Comment by johanbook** [comment]
I ran into this issue now as well when developing a monorepo with different pipelines triggering depending on which parts of the code is touched. It would be useful to be able to configure checks as required only if a pipeline runs

**Comment by ValentinGurkov** [comment]
I want to voice my frustration as well. I don't want to run my E2E tests and other code quality checks if I've updated the documentation. 
The GitHub documentation that states that the status checks should not be mandatory in this case is absurd, in my opinion. They should be mandatory **based on a given set of conditions**!

**Comment by mistercrunch** [comment]
I wanted to bring up the fact that jobs that use the `matrix` strategy are a bigger headache, where there's no way to add a check on the whole matrix, the `required_status_checks` have to be defined on the atomic instantiation of the matrix. I'd like to be able to depend on the whole matrix, and not have to list out every permutation. The "no-op" dummy job approach works, but it's a hack AND I have to duplicate my matrix semantics in that job and keep it in sync.

**Comment by T-Klug** [comment]
The fact you can't just say, "Any checks running need to pass before merging," is bonkers...

It is even worse that this is still an issue and that the accepted solutions are hacks or third parties taking over the GitHub action workflow.  

<img src="https://media3.giphy.com/media/l4EpciZRNKNrhVKpi/giphy.gif"/>

**Reply by mistercrunch** [reply]
At the very least referencing the matrix should work, and wildcards even better

**Reply by markdemich-fl** [reply]
Yes, looking at the UI of the PR, you can visually see that one of the checks run didn't pass.  Why can't they have a setting that does the same.  Seems odd that isn't built in.

**Reply by ghost** [reply]
> The fact you can't just say, "Any checks running need to pass before merging," is bonkers...

Exactly this. GitHub, please - I'm sure someone will chime and say how this breaks their workflow somehow, but the behavior really ought to be "any checks that run (whether conditionally or always) need to pass", not "this check must succeed for the PR to be merged" without any consideration if it's even appropriate to run or not. This leads to a lot of extra steps and workarounds in workflows to get right. And, please, make it one of the few features you bring to GHES...

**Reply by TWiStErRob** [reply]
Reading this again now, I think there's one issue with our logic: how would GitHub know *when* to check for green? For example imagine a workflow that triggers an external event and that creates a status check, or a `workflow_run` triggered job that runs after another workflow completed. With required checks we need to be explicit about what we're waiting for. With what we're requesting in this thread GitHub would need to be clairvoyant on the jobs to wait for. It could check the `pull_request` triggered workflows passed and auto-merge the PR, and then `workflow_run` or other status check might just fail *after*.

**Reply by robmaas** [reply]
> Reading this again now, I think there's one issue with our logic: how would GitHub know _when_ to check for green? For example imagine a workflow that triggers an external event and that creates a status check, or a `workflow_run` triggered job that runs after another workflow completed. With required checks we need to be explicit about what we're waiting for. With what we're requesting in this thread GitHub would need to be clairvoyant on the jobs to wait for. It could check the `pull_request` triggered workflows passed and auto-merge the PR, and then `workflow_run` or other status check might just fail _after_.

Just take a look at how `merge-gatekeeper` handles this. It's not rocket-science..

Edit: It would even be a viable solution to be able to list all required checks, but they would pass if the check would be `skipped`.
For example: Have 10 PR pipelines, most of them will be skipped and marked as such in the checks. If I would mark them as required the PR would not pass the way Github works currently.

**Reply by ghost** [reply]
> It could check the `pull_request` triggered workflows passed and auto-merge the PR, and then `workflow_run` or other status check might just fail _after_.

How is that different than the behavior today after the PR is merged?
